packages/zoltan2/test/helpers/AdapterForTests.hpp

```diff
@@ -120,10 +116,32 @@ class AdapterForTests{
    *
    * \return Ptr to the constructed adapter cast to the base class
    */
-  static AdapterWithOptionalCoordinateAdapter getAdapterForInput(
+  AdapterFactory(
     UserInputForTests *uinput, const ParameterList &pList,
     const RCP<const Comm<int> > &comm);
-  
+
+  ~AdapterFactory(); // handles deleting BaseAdapterRoot * data for adapter
+
+  AdapterWithOptionalCoordinateAdapter adaptersSet;
+
+  /*! \brief dynamic_cast adapter and then call getLocalNumIDs
+   */
+  size_t getLocalNumIDs() const {
+      #define GET_LOCAL_NUM_IDS(adapterClass)                                  \
+          return dynamic_cast<adapterClass*>(adaptersSet.main.adapter)->getLocalNumIDs();
+      Z2_TEST_UPCAST(adaptersSet.main.adapterType, GET_LOCAL_NUM_IDS);
+      throw std::logic_error( "getLocalNumIDs() failed to match adapter name" );
+  }
+
+  /*! \brief dynamic_cast adapter and then call getIDsView
+   */
+  void getIDsView(const zgno_t *&Ids) const {
```

Comment:
I made one further change and moved these factory 'utility' functions into test driver or ComparisonHelper as needed. That makes the factories cleaner. Still leaves the question of how to best handle the conversions (the places with Z2_TEST_UPCAST) and I would like to improve it. A few methods have no reference to the adapter such as Problem::solve() or BaseAdapter::getLocalNumIDs() so these can be called straight from the non-templated root class without any need to upcast. I demonstrated this in the current PR but would like to discuss this further as that may not be desirable for layout and could be switched back easily.